packages/suite/src/middlewares/firmware/firmwareMiddleware.ts

```diff
@@ -1,96 +1,99 @@
-import { SUITE } from 'src/actions/suite/constants';
-import { firmwareActions, validateFirmwareHash } from 'src/actions/firmware/firmwareActions';
-import { firmwareUpdate } from 'src/actions/firmware/firmwareThunks';
-import type { MiddlewareAPI } from 'redux';
-import type { AppState, Action, Dispatch } from 'src/types/suite';
+import { isAnyOf } from '@reduxjs/toolkit';
 
+import { createMiddlewareWithExtraDeps } from '@suite-common/redux-utils';
+
+import { firmwareActions } from 'src/actions/firmware/firmwareActions';
+import { validateFirmwareHash, firmwareUpdate } from 'src/actions/firmware/firmwareThunks';
 import { selectFirmware } from 'src/reducers/firmware/firmwareReducer';
 
-const firmwareMiddleware =
-    (api: MiddlewareAPI<Dispatch, AppState>) =>
-    (next: Dispatch) =>
-    (action: Action): Action => {
-        const firmware = selectFirmware(api.getState());
-        const { router } = api.getState();
+export const prepareFirmwareMiddleware = createMiddlewareWithExtraDeps(
+    (action, { getState, dispatch, extra, next }) => {
+        const {
+            selectors: { selectDevice, selectRouterApp },
+            actions: { appChanged, setSelectedDevice, updateSelectedDevice },
+        } = extra;
+        const firmware = selectFirmware(getState());
+        const device = selectDevice(getState());
+        const routerApp = selectRouterApp(getState());
         const { prevDevice } = firmware;
+
         // pass action
         next(action);
 
-        const { status, intermediaryInstalled } = api.getState().firmware;
+        const { status, intermediaryInstalled } = firmware;
+
+        if (isAnyOf(appChanged)(action)) {
```

Comment:
nitpick: simply `match` can be used instead of isAnyOf for single action type?

Reply:
[1b917a5](https://github.com/trezor/trezor-suite/pull/9141/commits/1b917a5b644ba54290dc2e6f8e8dffe3214eeb95)